zebra-consensus/src/block/tests.rs

```diff
@@ -10,9 +10,74 @@ use tower::{util::ServiceExt, Service};
 
 use zebra_chain::block::Block;
 use zebra_chain::block::BlockHeader;
-use zebra_chain::serialization::ZcashDeserialize;
+use zebra_chain::serialization::{ZcashDeserialize, ZcashDeserializeInto};
 use zebra_chain::transaction::Transaction;
 
+mod utils {
+
+    use super::*;
+
+    pub fn get_a_block() -> Block {
+        zebra_test::vectors::BLOCK_MAINNET_415000_BYTES
+            .zcash_deserialize_into::<Block>()
+            .unwrap()
+    }
+
+    pub fn get_a_prepared_block() -> (Arc<Block>, BlockHeaderHash) {
+        let block = Arc::new(get_a_block());
+        let hash: BlockHeaderHash = block.as_ref().into();
+
+        (block, hash)
+    }
+
+    #[spandoc::spandoc]
+    pub async fn verify<S, B>(
+        mut state_service: S,
+        mut block_verifier: B,
+        block: &Arc<Block>,
+        hash: &BlockHeaderHash,
+        verify_block: bool,
```

Comment:
using bools here makes it really confusing to infer what the test is doing. At a minimum we should probably be using `enum`s here so that the different modes are named at the call-site.

TIOLI: Even better might be a builder pattern type for constructing a test runner with a `verify().await` at the end to eventually call this `verify` fn. The fun part is coming up with the API. I haven't thought about it a ton but it might be fun for you to do so ill just give you a rough sketch and then you can run with it however you like.

What I'm imagining is something like this

```rust
use zebra_test::utils::Verify;
let block = zebra_test::utils::get_a_block();

zebra_test::utils::Verifier::new(state, verifier)
    .verify(block.and_then(Verify::Block).and_then(Verify::State))
    .await
    .expect("Ok")
    .verify(block.and_then(Verify::State))
    .await
    .expect_err("Repeated block should fail");
```

if you're confused about the `and_then` in the `verify` call it's a pattern I also use in color-eyre for constructing custom sections, so you can checkout the source [here](https://docs.rs/color-eyre/0.5.0/src/color_eyre/section/mod.rs.html#82-112) to see how to write code like that.

@hdevalence is also really good at this sort of API design so he might be a good person to talk to about ideas for this.

Reply:
Looking at this code it seems like what it's trying to do is make a transcript of the expected request / responses for the services it's testing, then check that either the state service or the block verifier are behaving correctly.

Instead of writing a specialized function to do this, it might be better to try to use the Transcript from the testing utilities: https://doc.zebra.zfnd.org/zebra_test/transcript/struct.Transcript.html

This isn't well-documented yet (sorry!) but the idea is that, as a Transcript user, you create a list of the expected requests / responses, and the Transcript can then check a given Service against the list you created.   There's an example of this in zebra-state/tests/basic.rs, which uses a `Lazy<Vec<(Request, Response)>>` to build up the transcript list.  Doing it this way would also eliminate the need for the `get_*_block` methods above.

One gap in the Transcript API is that the transcript only contains successful Responses, so it can't yet be used for checking that some request causes an error. This could probably be fixed but I'm not sure what the right way to check the errors is. (I will make a separate issue for that).

Reply:
https://github.com/ZcashFoundation/zebra/issues/676

Reply:
The method from @yaahc seems doable to me, not sure how much time it will take. In the other hand the @hdevalence seems to be even better but harder(for me to understand it) so i will need probably even more time on it. It also seems that the transcript api will need to be fixed first. I will of course will like to do it that way if this is the optimal. In the meantime i replaced the bool with enum at https://github.com/ZcashFoundation/zebra/pull/675/commits/e48aad55383a73ead2ec0cc6021a0c1ba741358c

Reply:
I've mostly finished the process of fixing the transcript in #678. My hope is we can get that merged tonight or tomorrow then you can go ahead and follow henry's suggestion for basing the tests on `Transcripts`, which is definitely what we should prefer.

Reply:
I don't think that changing the transcript API has to be a prerequisite to this, the current API works for checking that certain requests give certain responses, it just can't handle checking that things expected to error really do error.

Reply:
Ok good, i will start working on this. Thank you both.

Reply:
Actually i think https://github.com/ZcashFoundation/zebra/pull/678 is needed.